docs/useRegisterMutationMiddleware.md

```diff
@@ -0,0 +1,172 @@
+---
+layout: default
+title: "useRegisterMutationMiddleware"
+---
+
+# `useRegisterMutationMiddleware`
+
+React-admin lets you hook into the save logic of the forms in Creation and Edition pages using middleware functions. These functions "wrap" the main mutation (`dataProvider.create()` in a Creation page, `dataProvider.update()` in an Edition page), so you can add you own code to be executed before and after it. This allows you to perform various advanced form use cases, such as:
+
+- transforming the data passed to the main mutation,
+- updating the mutation parameters before it is called,
+- creating, updating or deleting related data,
+- adding performances logs,
+- etc.
+
+Middleware functions have access to the same parameters as the underlying mutation (`create` or `update`), and to a `next` function to call the next function in the mutation lifecycle.
+
+`useRegisterMutationMiddleware` allows to register a mutation middleware function for the current form.
+
+## Usage
+
+Define a middleware function, then use the hook to register it. 
+
+For example, a middleware for the create mutation looks like the following:
+
+```tsx
+import * as React from 'react';
+import {
+    useRegisterMutationMiddleware,
+    CreateParams,
+    MutateOptions,
+    CreateMutationFunction
+} from 'react-admin';
+
+const MyComponent = () => {
+    const createMiddleware = (
+        resource: string,
+        params: CreateParams,
+        options: MutateOptions,
+        next: CreateMutationFunction
+    ) => {
+        // Do something before the mutation
+
+        // Call the next middleware
+        next(resource, params, options);
+
+        // Do something after the mutation
+    }
```

Comment:
Shouldn't this be written as an async function? :thinking: 

```suggestion
    const createMiddleware = async (
        resource: string,
        params: CreateParams,
        options: MutateOptions,
        next: CreateMutationFunction
    ) => {
        // Do something before the mutation

        // Call the next middleware
        await next(resource, params, options);

        // Do something after the mutation
    }
```

Reply:
indeed